### ISSUE TITLE
webdav timeout: bug and solution

### ISSUE THREAD

**Issue by alpapan** [galaxyproject/galaxy#17372]
**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 23.1.5.dev0
Commit: 1fe37b099f6f7da20c2483157bb2c1d94a766c67

**Describe the bug**

I've setup my own Nextcloud and I've successfully made Galaxy to upload files.
Out of the box, it works with small files but it fails with large files.

Error log at end of this report

As it took me a few hours, I'm cataloguing here in case it's reproducible and useful.
I don't have a Galaxy solution as I had to edit a python lib file

*Solution part 1*
First I added the following to my Apache configurations:

*For Nextcloud*
```
Timeout 1200
```

*For Galaxy*

Under virtualhost
```
        ProxyTimeout 1200
        Timeout 1200
```

then for each ProxyPass / ProxyPassReverse added ` connectiontimeout=600` as an option (NB can only have 1 option, hence timeout is global)

FWIW, initially i had half the values above but sometimes i'd still get an error due to net traffic.

The above is necessary but insufficient because then python would fail

*Solution part 2*

It seems that there is a hardcoded time out when communicating via webdav.
I hoped that the `timeout` option in the webdav `file_sources_conf.yml` would be used but it is not used (yet).

```
- type: webdav
  id: nextcloud
  label: Some Label
  doc: Some Instructions 
  writable: true
  url: "my url"
  root: "/remote.php/webdav"
  login: ${user.preferences['nextcloud|username']}
  password: ${user.preferences['nextcloud|password']}
  timeout: 600
```

Note:
- `writable:true` was absent in the sample config; perhaps it can be added even as `writable:false`?
- `timeout` is not used

The way I solved it is by changing the default timeout in the __init__ of `webdavfs` library from None to 1200 (600 worked but had to increase).

Before that I had tried the following change but it didn't seem to be used for sending files.

```
--- a/lib/galaxy/util/__init__.py
+++ b/lib/galaxy/util/__init__.py
@@ -116,7 +116,7 @@ CHUNK_SIZE = 65536  # 64k
 DATABASE_MAX_STRING_SIZE = 32768
 DATABASE_MAX_STRING_SIZE_PRETTY = "32K"

-DEFAULT_SOCKET_TIMEOUT = 600
+DEFAULT_SOCKET_TIMEOUT = 1200

 gzip_magic = b"\x1f\x8b"
 bz2_magic = b"BZh"
@@ -1789,7 +1789,7 @@ def check_github_api_response_rate_limit(response):
         raise Exception(message)


-def download_to_file(url, dest_file_path, timeout=30, chunk_size=2**20):
+def download_to_file(url, dest_file_path, timeout=120, chunk_size=2**20):
     """Download a URL to a file in chunks."""
     with requests.get(url, timeout=timeout, stream=True) as r, open(dest_file_path, "wb") as f:
         for chunk in r.iter_content(chunk_size):
```

*I think it would be awesome if could use the `timeout` option in the webdav file_sources_conf.yml and pass it on to webdavs*

(and perhaps we could document the apache confs?)

let me know how i can assist.

**To Reproduce**
1. Setup a WebDav 
2. upload a 20 Gb BAM file

**Expected behavior**
FIle uploads with no error

**Error log**
```
Traceback (most recent call last):
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/fs/base.py", line 1417, in upload
    tools.copy_file_data(file, dst_file, chunk_size=chunk_size)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/fs/tools.py", line 51, in copy_file_data
    for chunk in iter(
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/fs/tools.py", line 52, in <lambda>
    lambda: read(_chunk_size) or None, None
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/apps/others/galaxy/tools/data_export/export_remote.py", line 79, in <module>
    main()
  File "/apps/others/galaxy/tools/data_export/export_remote.py", line 56, in main
    if write_if_not_exists(file_sources, target_uri, real_data_path):
  File "/apps/others/galaxy/tools/data_export/export_remote.py", line 35, in write_if_not_exists
    file_source.write_from(file_source_path.path, real_data_path)
  File "/apps/others/galaxy/lib/galaxy/files/sources/__init__.py", line 300, in write_from
    self._write_from(target_path, native_path, user_context=user_context, opts=opts)
  File "/apps/others/galaxy/lib/galaxy/files/sources/_pyfilesystem2.py", line 69, in _write_from
    openfs.upload(target_path, read_file)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/fs/base.py", line 1417, in upload
    tools.copy_file_data(file, dst_file, chunk_size=chunk_size)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdavfs/webdavfs.py", line 90, in close
    self.flush()
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdavfs/webdavfs.py", line 98, in flush
    self.res.read_from(self.data)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 989, in read_from
    self.client.upload_to(buff=buff, remote_path=self.urn.path())
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 67, in _wrapper
    res = fn(self, *args, **kw)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 541, in upload_to
    self.execute_request(action='upload', path=urn.quote(), data=buff)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 209, in execute_request
    response = self.session.request(
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 714, in urlopen
    httplib_response = self._make_request(
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 415, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connection.py", line 244, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/usr/lib/python3.8/http/client.py", line 1256, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1302, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1251, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1050, in _send_output
    self.send(chunk)
  File "/usr/lib/python3.8/http/client.py", line 972, in send
    self.sock.sendall(data)
 File "/usr/lib/python3.8/ssl.py", line 1233, in sendall
    v = self.send(byte_view[count:])
  File "/usr/lib/python3.8/ssl.py", line 1202, in send
    return self._sslobj.write(data)
```

I have an archive of the job and I can send via email on demand.

**Comment by nuwang** [comment]
Thanks for reporting this issue. Can you check the version of your fs.webdavfs package, it might be out-of-date? Support for specifying the timeout was added in: https://github.com/PyFilesystem/webdavfs/pull/33 so specifying the timeout in `file_sources_conf.yml` file should work.

**Comment by alpapan** [comment]
The requirement for `23.1.5.dev0` is `fs.webdavfs>=0.4.3`

My webdavfs is 
```
pip show fs.webdavfs
Name: fs.webdavfs
Version: 0.4.3.1
Summary: WebDAV support for pyfilesystem2
Home-page: http://pypi.python.org/pypi/fs.webdavfs/
Author: Andreas Jung and others
Author-email: info@zopyx.com
License: MIT
Location: /apps/others/galaxy/.venv/lib/python3.8/site-packages
Requires: cachetools, fs, python-dateutil, webdavclient3
Required-by:
```

**Comment by alpapan** [comment]
Sorry, the requirement for 23.1.5.dev0 is fs.webdavfs>=0.4.2

I made it 0.4.3 when troubleshooting

**Comment by alpapan** [comment]
Found the more complete error log, putting here FYI


```

Traceback (most recent call last):
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 714, in urlopen
    httplib_response = self._make_request(
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 415, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connection.py", line 244, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/usr/lib/python3.8/http/client.py", line 1256, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1302, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1251, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1050, in _send_output
    self.send(chunk)
  File "/usr/lib/python3.8/http/client.py", line 972, in send
    self.sock.sendall(data)
  File "/usr/lib/python3.8/ssl.py", line 1233, in sendall
    v = self.send(byte_view[count:])
  File "/usr/lib/python3.8/ssl.py", line 1202, in send
    return self._sslobj.write(data)
socket.timeout: The write operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 798, in urlopen
    retries = retries.increment(
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/util/retry.py", line 550, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/packages/six.py", line 769, in reraise
    raise value.with_traceback(tb)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 714, in urlopen
    httplib_response = self._make_request(
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 415, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/urllib3/connection.py", line 244, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/usr/lib/python3.8/http/client.py", line 1256, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1302, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1251, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1050, in _send_output
    self.send(chunk)
  File "/usr/lib/python3.8/http/client.py", line 972, in send
    self.sock.sendall(data)
  File "/usr/lib/python3.8/ssl.py", line 1233, in sendall
    v = self.send(byte_view[count:])
  File "/usr/lib/python3.8/ssl.py", line 1202, in send
    return self._sslobj.write(data)
urllib3.exceptions.ProtocolError: ('Connection aborted.', timeout('The write operation timed out'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 67, in _wrapper
    res = fn(self, *args, **kw)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 541, in upload_to
    self.execute_request(action='upload', path=urn.quote(), data=buff)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 209, in execute_request
    response = self.session.request(
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/requests/adapters.py", line 501, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', timeout('The write operation timed out'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/apps/others/galaxy/tools/data_export/export_remote.py", line 79, in <module>
    main()
  File "/apps/others/galaxy/tools/data_export/export_remote.py", line 56, in main
    if write_if_not_exists(file_sources, target_uri, real_data_path):
  File "/apps/others/galaxy/tools/data_export/export_remote.py", line 35, in write_if_not_exists
    file_source.write_from(file_source_path.path, real_data_path)
  File "/apps/others/galaxy/lib/galaxy/files/sources/__init__.py", line 300, in write_from
    self._write_from(target_path, native_path, user_context=user_context, opts=opts)
  File "/apps/others/galaxy/lib/galaxy/files/sources/_pyfilesystem2.py", line 69, in _write_from
    openfs.upload(target_path, read_file)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/fs/base.py", line 1417, in upload
    tools.copy_file_data(file, dst_file, chunk_size=chunk_size)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdavfs/webdavfs.py", line 91, in close
    super(WebDAVFile, self).close()
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdavfs/webdavfs.py", line 98, in flush
    self.res.read_from(self.data)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 989, in read_from
    self.client.upload_to(buff=buff, remote_path=self.urn.path())
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 69, in _wrapper
    raise NoConnection(self.webdav.hostname)
webdav3.exceptions.NoConnection: No connection with https://nextcloud.example.com

```

**Comment by alpapan** [comment]
it's a localhost webdav filesystem so it would be limited by the 1 gigabit connectivity

**Comment by alpapan** [comment]
my constructor had None and it caused the issue. when i put it to 600 it fixed it.
i don't have an issue anymore so it was definitely that...

**Comment by nuwang** [comment]
Any extra properties you add in `file_source_conf.xml` (including timeout), should be passed to the constructor as is on this line: https://github.com/galaxyproject/galaxy/blob/5c6dd8eb4be9aaaed9c03fbd960eff2f1ad03b87/lib/galaxy/files/sources/webdav.py#L26

If you can check whether it is indeed getting passed in (print out the props dictionary as well as the timeout value received by the WebDAVFS constructor), that might further help to narrow down the issue.

**Comment by alpapan** [comment]
btw @nuwang , is there a reason why it loads the whole file into memory?

I swear i saw a ticket somewhere that there was a python library update by the galaxy team that addressed that...



![image](https://github.com/galaxyproject/galaxy/assets/608484/cb312fd3-c8e9-44cb-92a0-2105fdb77c85)

**Comment by nuwang** [comment]
This was another PR on webdavfs: https://github.com/PyFilesystem/webdavfs/pull/34

**Comment by alpapan** [comment]
pip tells me i have 0.4.3.1 which is post july
last stable is 0.4.2

the changes you pointed out are in the `~/.venv/lib/python3.8/site-packages/webdavfs`


```
"metadata": {
        "metadata_version": "2.1",
        "name": "fs.webdavfs",
        "version": "0.4.3.1",
        "platform": [
          "any"
        ],
```

**Comment by alpapan** [comment]
> Since the connection speeds are pretty good in Galaxy Australia, I'm assuming that slow connectivity is a factor in your case. 

tangent:
FWIW, when I use the NFS server, the throughput is averaging 354.10 Mb/s on the write side which should take 90 sec per 32g bam file.
However, each 32gb file takes about 6 min to write. 

Comparing time logs between host running export_remote.py (which reads file in memory) and the webdav host, there is a 4 min difference (the exec host takes 4 min extra)....

**Comment by alpapan** [comment]
How can I tell if the script is using the right webdavfs library?

There is no way I can think of that it doesn't use the galaxy one but given evidence of the memory use and the fact it has the latest one installed under galaxy's .venv, I have no idea what could be causing it.

Assuming the memory issue is fixed that is.

**Comment by alpapan** [comment]
I tried it a dozen times looking for configurations that could be causing it. 

It always does after 5 minutes (300 seconds).

Even though file should have been transferred in less than 90 seconds.

It keeps failing to transfer the last few bytes and index.

I'm going to give up on webdavfs I think. Between this and the memory issue, it's just not worth it.

```
Traceback (most recent call last):
  File "/apps/others/galaxy/tools/data_export/export_remote.py", line 79, in <module>
    main()
  File "/apps/others/galaxy/tools/data_export/export_remote.py", line 56, in main
    if write_if_not_exists(file_sources, target_uri, real_data_path):
  File "/apps/others/galaxy/tools/data_export/export_remote.py", line 35, in write_if_not_exists
    file_source.write_from(file_source_path.path, real_data_path)
  File "/apps/others/galaxy/lib/galaxy/files/sources/__init__.py", line 300, in write_from
    self._write_from(target_path, native_path, user_context=user_context, opts=opts)
  File "/apps/others/galaxy/lib/galaxy/files/sources/_pyfilesystem2.py", line 69, in _write_from
    openfs.upload(target_path, read_file)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/fs/base.py", line 1417, in upload
    tools.copy_file_data(file, dst_file, chunk_size=chunk_size)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdavfs/webdavfs.py", line 91, in close
    super(WebDAVFile, self).close()
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdavfs/webdavfs.py", line 98, in flush
    self.res.read_from(self.data)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 989, in read_from
    self.client.upload_to(buff=buff, remote_path=self.urn.path())
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 67, in _wrapper
    res = fn(self, *args, **kw)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 541, in upload_to
    self.execute_request(action='upload', path=urn.quote(), data=buff)
  File "/apps/others/galaxy/.venv/lib/python3.8/site-packages/webdav3/client.py", line 230, in execute_request
    raise ResponseErrorCode(url=self.get_url(path), code=response.status_code, message=response.content)
webdav3.exceptions.ResponseErrorCode: Request to https://nextcloud.stressedfruitfly.com/remote.php/webdav/Galaxy/workflow_results/ResistantS3GraceOriginal%20vs%20MurcottPhoenix%20Alignment%20concordant.bam failed with code 408 and message: b'<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">\n<html><head>\n<title>408 Request Timeout</title>\n</head><body>\n<h1>Request Timeout</h1>\n<p>Server timeout waiting for the HTTP request from the client.</p>\n<hr>\n<address>Apache/2.4.58 (Ubuntu) Server at nextcloud.example.com Port 443</address>\n</body></html>\n'
```

**Comment by alpapan** [comment]
Nextcloud reports:

```
Expected filesize of 24585021362 bytes but read (from Nextcloud client) and wrote (to Nextcloud storage) 49152 bytes. Could either be a network problem on the sending side or a problem writing to the storage on the server side.
```

It does the same error four times in a row. It's not network congestion, it is a code issue definitely.

**Comment by nuwang** [comment]
You could try force upgrading the dependency from within Galaxy's venv as follows:
```
source .venv/bin/activate
pip install --upgrade git+https://github.com/PyFilesystem/webdavfs
```

**Comment by alpapan** [comment]
As mentioned, it was already updated. The latest version has been there since `Jan 26 17:39 __init__.py`

```
source .venv/bin/activate
pip install --upgrade git+https://github.com/PyFilesystem/webdavfs
Collecting git+https://github.com/PyFilesystem/webdavfs
  Cloning https://github.com/PyFilesystem/webdavfs to ./tmp/pip-req-build-l2t2goqt
  Running command git clone --filter=blob:none --quiet https://github.com/PyFilesystem/webdavfs /apps/others/galaxy/tmp/pip-req-build-l2t2goqt
  Resolved https://github.com/PyFilesystem/webdavfs to commit ab698904c8a4a07d63834f9ec5422819ed17d6e9
  Preparing metadata (setup.py) ... done
Requirement already satisfied: fs>2.0 in ./.venv/lib/python3.8/site-packages (from fs.webdavfs==0.4.3.1) (2.4.16)
Requirement already satisfied: webdavclient3 in ./.venv/lib/python3.8/site-packages (from fs.webdavfs==0.4.3.1) (3.14.6)
Requirement already satisfied: python-dateutil in ./.venv/lib/python3.8/site-packages (from fs.webdavfs==0.4.3.1) (2.8.2)
Requirement already satisfied: cachetools in ./.venv/lib/python3.8/site-packages (from fs.webdavfs==0.4.3.1) (3.1.0)
Requirement already satisfied: appdirs~=1.4.3 in ./.venv/lib/python3.8/site-packages (from fs>2.0->fs.webdavfs==0.4.3.1) (1.4.4)
Requirement already satisfied: setuptools in ./.venv/lib/python3.8/site-packages (from fs>2.0->fs.webdavfs==0.4.3.1) (67.8.0)
Requirement already satisfied: six~=1.10 in ./.venv/lib/python3.8/site-packages (from fs>2.0->fs.webdavfs==0.4.3.1) (1.16.0)
Requirement already satisfied: requests in ./.venv/lib/python3.8/site-packages (from webdavclient3->fs.webdavfs==0.4.3.1) (2.31.0)
Requirement already satisfied: lxml in ./.venv/lib/python3.8/site-packages (from webdavclient3->fs.webdavfs==0.4.3.1) (4.9.2)
Requirement already satisfied: charset-normalizer<4,>=2 in ./.venv/lib/python3.8/site-packages (from requests->webdavclient3->fs.webdavfs==0.4.3.1) (3.1.0)
Requirement already satisfied: idna<4,>=2.5 in ./.venv/lib/python3.8/site-packages (from requests->webdavclient3->fs.webdavfs==0.4.3.1) (3.4)
Requirement already satisfied: urllib3<3,>=1.21.1 in ./.venv/lib/python3.8/site-packages (from requests->webdavclient3->fs.webdavfs==0.4.3.1) (1.26.16)
Requirement already satisfied: certifi>=2017.4.17 in ./.venv/lib/python3.8/site-packages (from requests->webdavclient3->fs.webdavfs==0.4.3.1) (2023.5.7)
```

**Comment by nuwang** [comment]
One last thing to try:
```
source .venv/bin/activate
pip uninstall webdavfs
pip uninstall webdavclient3
pip install --upgrade git+https://github.com/ezhov-evgeny/webdav-client-python-3
pip install --upgrade git+https://github.com/PyFilesystem/webdavfs
```

**Comment by davelopez** [comment]
Wait, It could also be falling back to the in-memory file if `use_temp_files` and `temp_path` it is not specified in the config https://github.com/PyFilesystem/webdavfs/pull/34/files#diff-419f6744efd46a5bf1d476b1706a3fb410405ec3d2413c7025180c2405ab6239R167

file_sources_conf.yml
```
- type: webdav
  ...
  use_temp_files: true
  temp_path: /path/to/temp
  ...
```

**Comment by alpapan** [comment]
> Wait, It could also be falling back to the in-memory file if `use_temp_files` and `temp_path` it is not specified in the config https://github.com/PyFilesystem/webdavfs/pull/34/files#diff-419f6744efd46a5bf1d476b1706a3fb410405ec3d2413c7025180c2405ab6239R167
> 
> file_sources_conf.yml
> 
> ```
> - type: webdav
>   ...
>   use_temp_files: true
>   temp_path: /path/to/temp
>   ...
> ```

that will fix the memory issue thanks! could we perhaps add it to the sample config?

**Comment by davelopez** [comment]
Yeah definitely! my bad I forgot to add it to the sample at the time, sorry :sweat: 

I will open a PR in a few, thanks for reporting and sorting out the issue!

**Comment by alpapan** [comment]
> Wait, It could also be falling back to the in-memory file if `use_temp_files` and `temp_path` it is not specified in the config https://github.com/PyFilesystem/webdavfs/pull/34/files#diff-419f6744efd46a5bf1d476b1706a3fb410405ec3d2413c7025180c2405ab6239R167
> 
> file_sources_conf.yml
> 
> ```
> - type: webdav
>   ...
>   use_temp_files: true
>   temp_path: /path/to/temp
>   ...
> ```

confirming this did fix the memory issue and the timeouts.

**Comment by alpapan** [comment]
> Thanks for reporting this issue. Can you check the version of your fs.webdavfs package, it might be out-of-date? Support for specifying the timeout was added in: [PyFilesystem/webdavfs#33](https://github.com/PyFilesystem/webdavfs/pull/33) so specifying the timeout in `file_sources_conf.yml` file should work.

@nuwang fwiw, the aussie galaxy config file is not using temp file directive, so not sure how your server is able to not use so much memory. It should be falling back to in mem

**Comment by davelopez** [comment]
Since we require a temp path making it the default upstream might be a breaking change or controversial, but I do like the idea of specifying a default temp path in Galaxy's filesource implementation. We probably should do that :+1:

**Comment by alpapan** [comment]
please don't forget documenting writable or setting `writable:false`

**Comment by alpapan** [comment]
i meant if it was written there then stupid people like me will know it can be set to true... i know it sounds stupid to me now that i know how it works but when i first started using them i didn't know 

_btw, my timeouts are still there. i think i figured out when they are caused (bam files with index set to upload) but that's for another ticket, another day_

**Comment by davelopez** [comment]
Ahh I see, will add a comment for writable then :+1:
